### ISSUE TITLE
Accessing movie array results outside of search function

### ISSUE THREAD

**Issue by skoop12** [uchicago-mobi/MPCS51030-2019-Winter-Forum#30]
I can't seem to connect the movies results from the search function to the actual movies array connected to my table cells.  

When I call movieService.search in viewDidLoad, the function includes a print(movies.count) and it prints out 20 (so I know the array is not empty).  However, when I run my app, nothing shows in the table.  

I tried to debug by calling print(movies) outside of the function (in viewDidLoad and outside of it) and it prints an empty array. Somehow the movies output from the function isn't connecting to the movies array declared.  I thought thats what "guard let movies = movies" in the search completion field was supposed to be doing, but I guess not.  Is there another step needed to connect the output? 

 I have also tried putting reloadData() in different locations but that isn't resolving the issue either.

**Comment by hbennett766** [comment]
Leave `reloadData()` in that original spot, that placement is correct. Let's think about what's happening during a networking call. It is both confusing and important!

When `viewDidLoad()` is called, the table view has already populated with the empty data. We make our networking call, `search`, and via a completion block, tell the code inside the block "Hey, don't execute until you hear back that this networking has been done."

Once the search results have come back from the cloud or wherever, we call back and say "Hey completion block, do your thing. Here is what I got back from the internet."

So here we are with some response. Is it an error, is it the movie json, is it nothing? We have to find out! 🕵🏼‍♀️ The line that does that sleuthing is `guard let movies = movies, error == nil else { // print error and return early }`

In English, that says "I need this variable `movies` to exist and be assigned to a non-optional variable that is also called `movies`. I _also_ need `error` to be equal to `nil`. If this is not the case, continue into the `else` block, handle the failure, and return early." Because we can't repopulate a table with no data.

Side note: The first `movies` in that guard statement is the new non-optional variable you are creating. You can call it whatever, but it's standard to keep the name of the original variable. The second `movies` is the optional value coming in from the server.

In this scenario, we now have a non-optional `movies` array, but it only persists _within this method_, because it was declared as an instance variable. We need to copy that array to somewhere that it will persist throughout the life of the view controller.

Enter your class-level declaration of the `movies` array. Where can you assign your instance level `movies` array to your class level `movies` array?

**Comment by hbennett766** [comment]
More info on [guard, if-let, and defer statements](https://medium.com/@abhimuralidharan/if-let-if-var-guard-let-and-defer-statements-in-swift-4f87fe857eb6) if you're interested. Lots of ways to portray conditionals in Swift. Up to you to determine the best way given the situation.

**Comment by skoop12** [comment]
Ok I think I understand what you're saying, but I still can't get it to work... so maybe not.  I tried to assign my class level variable movies to the movies created in the search function, but I'm still getting 0 movies when I try to test it outside of the function.  

I added the line `self.movies = movies` inside the function after print(Movies fetched...), but it's still not working outside of the function. Is there another way to assign class level variables in swift?

**Comment by hbennett766** [comment]
I pulled down your repository and checked out what was going on, thanks for keeping that up to date 👍🏻 Your assignment of the class-level `movies` array is correct.

When I said "leave `tableView.reloadData()` where it is", I made a bad assumption about where it was being called in the first place. That's what's causing you issues. Sorry about that initial statement, assumptions tend to cause unexpected results in programming...

Inside `viewDidLoad()`, only the code _within_ the completion block will be held up until we receive data back from the network. Everything outside of a completion block will be run asynchronously and immediately.

When in the lifecycle is `tableView.reloadData()` being called right now, and where does that call need to live to make sure the class-level array is populated beforehand?
